### ISSUE TITLE
light-the-torch future maintenance discussion

### ISSUE THREAD

**Issue by jamesobutler** [Slicer/light-the-torch#152]
Conversation originally pulled from #151:

@pmeier:

> To give some context: I'm no longer a PyTorch dev / power user and thus have very little use for light-the-torch in my day-to-day. And with that maintaining this project became more of a burden to me. However, the fact that it still works and has roughly 1k downloads per week after being unmaintained for a good year, is a testament that this project shouldn't be archived.
> 
> Since you basically did all the maintainer work in this PR, would you like to have maintainer rights on this repository so that you can push fixes yourself in the future?
> 

@jamesobutler:

> > Since you basically did all the maintainer work in this PR, would you like to have maintainer rights on this repository so that you can push fixes yourself in the future?
> 
> Yes, that would be great!
> 
> > To give some context: I'm no longer a PyTorch dev / power user and thus have very little use for light-the-torch in my day-to-day. And with that maintaining this project became more of a burden to me. 
> 
> I can discuss with the other Slicer main developers (e.g. @lassoan, also a `light-the-torch` contributor, see [here](https://github.com/pmeier/light-the-torch/commits/main/?author=lassoan)), but would you potentially consider [transferring](https://docs.github.com/en/repositories/creating-and-managing-repositories/transferring-a-repository) the repo to the https://github.com/Slicer GitHub organization (or to me to then transfer to the Slicer org)? Any links to https://github.com/pmeier/light-the-torch would automatically redirect to a migrated https://github.com/Slicer/light-the-torch location. Then myself and other Slicer main developers could continue maintenance of the project. There are probably several of us that would like to keep this going for the Slicer community. We would additionally add props to you in the README with a history section indicating the work was primarily under your account and that it was later transferred to the Slicer organization to carry maintenance forward. If you'd like to keep this repo under your GitHub name, I totally understand. Only brought this up if this repo has become a burden to the point you want to pass it to another person/organization. Otherwise we can also work within the bounds of getting write access to this repo and get access for other potential maintainers from the Slicer community. 

@pmeier:

> > but would you potentially consider [transferring](https://docs.github.com/en/repositories/creating-and-managing-repositories/transferring-a-repository) the repo to the https://github.com/Slicer GitHub organization (or to me to then transfer to the Slicer org)?
> 
> I'm not exactly sure why, that hit hard reading it the first time. I didn't like it at all. But the longer I thought about it, the more sense it makes to me. I wanted to have this PR finished by no later than Thursday and here I am sitting on a Saturday night, because I simply didn't find any time earlier. Coupled with the fact that I'm no longer a user of this project, it makes little sense to keep me as maintainer and by proxy to keep this as my repository. So, yes, I'm happy to transfer it anywhere of your choosing. I'd prefer under the Slicer umbrella to make this a proper community project, but I'm ok with either.

**Comment by jamesobutler** [comment]
Apologies if the suggestion of migrating the project to a new home was undesired and unexpected! I see you take pride in having this project under your name with your [mention](https://github.com/pmeier/pmeier/blob/3879e001ce289ac117609e295c414ceeaebe7fd3/README.md?plain=1#L12) of it in your personal GitHub readme. The suggestion to transfer the repository was purely from the perspective of trying to reduce any burden that you stated this project had become.

Using the [SlicerDeveloperToolsForExtension](https://github.com/Slicer/SlicerDeveloperToolsForExtensions) and its extension stats module, I was able to check how many times the [SlicerPyTorch](https://github.com/fepegar/SlicerPyTorch) extension was downloaded where this extension utilizes your `light-the-torch` package. The [Slicer](https://slicer.org/) community has a latest stable release and preview builds until the next stable. The current Slicer stable version 5.6.2 and Slicer Preview post-5.6.2 (5.7.0.YYYY-MM-DD) have combined for about 26,000 downloads in the past 8 months or so. 
![{112D8196-4577-4C45-8846-94F6965B9EB1}](https://github.com/user-attachments/assets/0c8dab6f-198c-4a10-913d-7313b56a741d)
Assuming every time someone downloads the SlicerPyTorch extension they ultimately install torch, that would put `light-the-torch` download usage to about 3250 downloads a month or maybe ~800 a week. With your statement about 1k weekly downloads for `light-the-torch`, I would assume that the large majority originates from the Slicer communities usage of the SlicerPyTorch extension. Currently there are 14 extensions in the Slicer [Extensions Index](https://github.com/Slicer/ExtensionsIndex) that have SlicerPyTorch as an extension and this has definitely been growing in the community over the past year. This is why the Slicer community definitely has a vested interest in the health of this project 🎉 . What originally triggered me to issue #151, was some discussion in a Slicer extension review process where the developer described `light-the-torch` was only installing CUDA compute 11.8 (since no cu121, cu124 support in `light-the-torch` v0.7.0) and the developer was describing that their code seemingly didn't have issues when running on newer CUDA 12.x compute. This has ultimately led to some discussions about how to handle proceeding with the extension and waiting on a `light-the-torch` update or to create workarounds. See [this PR](https://github.com/Slicer/ExtensionsIndex/pull/2127#issuecomment-2537354178) if you're curious.

I was re-reviewing GitHub [repository roles](https://docs.github.com/en/organizations/managing-user-access-to-your-organizations-repositories/managing-repository-roles/repository-roles-for-an-organization#permissions-for-each-role), and the maintainer role to manage issues, pull requests and create releases. One thing it doesn't provide is the ability to manage individual and team access to the repository or manage branch protection rules and repository rulesets. This seem like important things to come up if I wanted another Slicer developer to help with the maintenance and have these elevated privileges that wouldn't come with just being an outside contributor.

Questions:
- Would you be available and provide timely responses for things such as request for new maintainer or a change in branch protection rules (aka enforcing new status checks or creating new protected branches for possible maintenance branches)?
  - Alternatively, you could provide an active developer to the project as a trusted secondary admin. This could be adjusted as needed in the future. This could also support proper maintenance while maintaining the repo under your GitHub username.  
- Are you comfortable lettings other developers publish a new release?
  - If yes, if the project is under your GitHub username, are you comfortable with others publishing a new release that has your name still tied to the project? We don't want to give you a bad reputation if we push a bad release or something. Not planning on it though haha! I guess the project could still be under your GitHub username with the additional of `maintainers` to the [metadata](https://packaging.python.org/en/latest/tutorials/packaging-projects/#configuring-metadata). Any additional ideas that can clearly reflect the project was originally created by you, though you may not have had any oversight or approval of a future release. 
  https://github.com/pmeier/light-the-torch/blob/19e77b794c1b4f64a8591a764348e259ddd4055d/pyproject.toml#L19-L21

**Comment by pmeier** [comment]
> Apologies if the suggestion of migrating the project to a new home was undesired and unexpected! I see you take pride in having this project under your name with your [mention](https://github.com/pmeier/pmeier/blob/3879e001ce289ac117609e295c414ceeaebe7fd3/README.md?plain=1#L12) of it in your personal GitHub readme. The suggestion to transfer the repository was purely from the perspective of trying to reduce any burden that you stated this project had become.

Indeed I take pride (for the lack of a better word) in it. This is why I was so hesitant _in the beginning_. I mulled this over all week and came to the decision that it is rather selfish and not in the spirit of OSS to hog this project when I don't have the time to maintain it anymore but others are willing to.

Meaning, there is no need anymore to convince me. Sorry if my original communication was not clear enough in this regard. And most important: thanks for keeping this project alive!

---

The reason I wanted to have this issue is to discuss the details of transferring the project over.

1. Who is going to be transferred to? You personally or the [Slicer](https://github.com/orgs/Slicer/people) organization? For the latter, I'd like to have confirmation of at least one more public member that they want this.
2. What needs to happen in the code? I currently see two things:
   1. Add a new maintainers key to the project metadata as you detailed above. Who should that be? Only you for now or someone else? If the latter, these people should please speak up in this thread.
   2. Change the license from
      https://github.com/pmeier/light-the-torch/blob/33397cbe45d07b51ad8ee76b004571a4c236e37f/LICENSE#L3

      to

      ```
      Copyright (c) 2020-2024, Philip Meier
      Copyright (c) 2024-, ???
      ```

      where `???` is up for decision
3. What needs to happen on the project level? Apart from transferring the repo from 1., I also need to transfer the project on PyPI. Again whoever wants to own it, organization or individual needs to be decided.

**Comment by jamesobutler** [comment]
> And most important: thanks for keeping this project alive!

Your contribution will forever live on wherever the repository is hosted. 🙏🏻 

I will ask the other core Slicer developers at the weekly Tuesday Slicer hangout meeting(or if @lassoan wants to already reply here). These meetings are announced with links to join in posts made at  https://discourse.slicer.org/c/community/12. 

I have the permissions to create a repository in the Slicer GitHub organization so ultimately if I am made an Admin of this repository then I would be able to complete the transfer process to the Slicer organization. If the other Slicer developers object to the transfer then we can keep it under your repository name for the time being. I’ll let post a summary of the Slicer hangout conversation here in this issue.

Thanks for pointing out those copyright items. I’ll discuss with the Slicer devs about best way to update. My preference would be to list the author as the “Slicer Community”, but if best practice to list specific people then we can list the primary maintainer (might be me).

**Comment by jamesobutler** [comment]
@pmeier The Slicer community is good with the migration of this repository to the Sicer GitHub organization. One comment that came up was a surprise that Slicer is indeed such a heavy majority of the users utilizing this `light-the-torch` and an interest about what all the other regular torch developers do to download the appropriate CUDA whl of torch for their platform. Does everyone else really just use the table at https://pytorch.org/ and hardcoded download links such as `pip3 install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu124`? Is this utility something that the PyTorch team would be interested in taking in for their GitHub organization? This would be a reason to not migrate to the Slicer GitHub organization. If PyTorch began utilizing this utility it would help increase the download usage which is currently ~1k weekly downloads (https://pypistats.org/packages/light-the-torch).

**Comment by pmeier** [comment]
> Does everyone else really just use the table at https://pytorch.org/ and hardcoded download links such as `pip3 install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu124`?

To my best knowledge, yes.

> Is this utility something that the PyTorch team would be interested in taking in for their GitHub organization?

I haven't talked to them in a while, but when I first created `light-the-torch`, there wasn't. Maybe there is some interest now, but I personally don't have the availability to push for it.

> The Slicer community is good with the migration of this repository to the Sicer GitHub organization.

Ok, that settles 1. of https://github.com/pmeier/light-the-torch/issues/152#issuecomment-2543937405. Maybe I'm being overly cautious here, but could any Slicer member other than @jamesobutler chime in to confirm this?

**Comment by jamesobutler** [comment]
@lassoan @jcfr would you be able to provide verbal approval here from our discussion at the Slicer hangout this morning?

**Comment by lassoan** [comment]
Chiming in! Sorry, I did not speak up because I everything looked good, I did not feel the need to step in. But yes, I confirm that we would welcome the light-the-torch project in the Slicer organization.

**Comment by jcfr** [comment]
Thanks @jamesobutler for coordinating this initiative :rocket: and for @pmeier for being so  supportive and responsive :pray:

To follow-up on @lassoan comment, I also support transferring the project to the Slicer organization :ok_hand:

**Comment by gtkacz** [comment]
Hey there! Sorry for chiming in but I noticed that the repository still doesn't belong to the Slicer organization, and the last activity on this repo seems to coincide with the date of @jcfr's last comment here, where @pmeier created a [PR](https://github.com/pmeier/light-the-torch/commits/prepare-transition/) and commit by the name of "prepare transition". Were there any hiccups? Is the project going to be maintained going forward? Thanks for your time!

**Comment by jamesobutler** [comment]
@pmeier I have accepted the invitation.

**Comment by pmeier** [comment]
Reading https://docs.github.com/en/account-and-profile/setting-up-and-managing-your-personal-account-on-github/managing-user-account-settings/permission-levels-for-a-personal-account-repository#collaborator-access-for-a-repository-owned-by-a-personal-account, it looks like I can't give you admin perms 😡 Let me try to transfer it to you personally and you move from there.

**Comment by pmeier** [comment]
Damn, this is frustrating

![Image](https://github.com/user-attachments/assets/b629b1b6-f215-46d6-b3ab-548b77f16f61)

I don't know if it will work after that, but you'll have to (re-)move your fork ...

**Comment by jamesobutler** [comment]
https://github.com/jamesobutler/light-the-torch fork has been deleted.

**Comment by pmeier** [comment]
Ok, that went through 🎉

**Comment by jamesobutler** [comment]
The transfer to https://github.com/jamesobutler/light-the-torch was complete and then I have completed the transfer to the Slicer organization at https://github.com/Slicer/light-the-torch as we can now see from this repository.

**Comment by pmeier** [comment]
Awesome. Thank you for being patient with me. The only thing left is making someone admin on PyPI. @jamesobutler do you have an account there?

**Comment by jamesobutler** [comment]
I have re-forked this repo so that https://github.com/jamesobutler/light-the-torch no longer redirects to https://github.com/Slicer/light-the-torch which was set up upon me transferring this repo from my account to the Slicer organization.

@pmeier Yes, I have a PyPI account. Can you find me at @jamesobutler ?

**Comment by pmeier** [comment]
Invite sent out.

**Comment by jamesobutler** [comment]
Invitation accepted. I will work with the other Slicer devs to obtain relevant administration of the project on PyPI and GitHub. Do you still have admin privileges for this current GitHub repo to make me admin as well?

**Comment by pmeier** [comment]
> Do you still have admin privileges for this current GitHub repo to make me admin as well?

I don't.

**Comment by lassoan** [comment]
@pmeier @jamesobutler both of you are admins now

**Comment by jamesobutler** [comment]
@lassoan Would you like to additionally have the "Owner" role of the `light-the-torch` PyPI package? This way it won't be a sole owner project.

**Comment by lassoan** [comment]
Yes, that's a good idea. My pypi username is lassoan.

**Comment by jamesobutler** [comment]
Invite has been sent.

**Comment by lassoan** [comment]
Thanks, accepted.

**Comment by pmeier** [comment]
Just FYI: I've un-admined myself. I'm happy to stay on as maintainer emeritus for context, but I won't do admin stuff any time soon. Thus, I also don't need the permissions. If the need arises for whatever reason, I can be given them then.
